### PR TITLE
CORE-5472: Decline registration if the member or MGM type is wrong

### PR DESCRIPTION
**Pull request by yift-r3**

### CODE REVIEW

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/DeclineRegistrationHandler.kt [hunk 1]
```diff
@@ -35,6 +39,12 @@ class DeclineRegistrationHandler(
         val declinedBy = state.mgm
         val declinedMember = state.registeringMember
         val registrationId = state.registrationId
+        if (memberTypeChecker.isMgm(declinedMember)) {
```
Comment (nikinagy) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/DeclineRegistrationHandler.kt:42]: Do we need to check here again? Would it be enough to check the reason in the Declined message?

Comment (nikinagy) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/DeclineRegistrationHandler.kt:42]: On a second thought maybe it is really better to check here too to be fool-proof.
Although if there was any message in DeclineRegistration command, can we log out here too?

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt [hunk 2]
```diff
@@ -15,39 +19,61 @@ import net.corda.membership.persistence.client.MembershipPersistenceClient
...
+                it.key == VERIFIED
+            }?.value?.toBooleanStrictOrNull() ?: false
+            if (!success) {
+                val reasons = command.verificationResponse.payload.items
+                    .filter { it.key == FAILURE_REASONS }
+                    .map { it.value }
+                val message = "Could no verify registration $registrationId - $reasons"
+                throw CordaRuntimeException(message)
+            }
+            if (memberTypeChecker.isMgm(member)) {
```
Comment (nikinagy) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt:58]: Is there a way to have these checks somewhere in a common helper? Might be easier to maintain and add additional checks in the future.

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/RegistrationProcessorTest.kt [hunk 3]
```diff
@@ -116,38 +118,32 @@ class RegistrationProcessorTest {
     lateinit var membershipPersistenceClient: MembershipPersistenceClient
     lateinit var membershipQueryClient: MembershipQueryClient
```
Comment (nikinagy) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/RegistrationProcessorTest.kt]: Can we make these as private too?

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/MemberTypeCheckerTest.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        on { lookup(knownMember) } doReturn memberInfo
+    }
+    private val membershipGroupReaderProvider = mock<MembershipGroupReaderProvider> {
+        on { getGroupReader(any()) } doReturn reader
+    }
+
+    private val memberTypeChecker = MemberTypeChecker(membershipGroupReaderProvider)
+
+    @Test
+    fun `isMgm return false for unknown member`() {
```
Comment (nikinagy) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/MemberTypeCheckerTest.kt:48]: can we have tests resulting in true too?

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandlerTest.kt [hunk 5]
```diff
@@ -278,11 +286,40 @@ class ApproveRegistrationHandlerTest {
...
+        val results = handler.invoke(state, key, command)
+
+        assertThat(results.outputStates)
+            .hasSize(1)
+            .allSatisfy {
+                assertThat(it.topic).isEqualTo(REGISTRATION_COMMAND_TOPIC)
+                val value = (it.value as? RegistrationCommand)?.command
+                assertThat(value)
+                    .isNotNull
+                    .isInstanceOf(DeclineRegistration::class.java)
```
Comment (nikinagy) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandlerTest.kt:302]: Maybe check the reason field is not empty?

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandlerTest.kt [hunk 6]
```diff
@@ -103,6 +119,68 @@ class ProcessMemberVerificationResponseHandlerTest {
...
+
+        verify(membershipPersistenceClient, never()).setRegistrationRequestStatus(
+            any(),
+            any(),
+            any(),
+        )
+
+        assertThat(result.outputStates).hasSize(1)
+            .anyMatch {
+                (it.value as? RegistrationCommand)?.command is DeclineRegistration
```
Comment (nikinagy) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandlerTest.kt]: Check the reason is not empty here too.

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandlerTest.kt [hunk 7]
```diff
@@ -96,24 +93,22 @@ class StartRegistrationHandlerTest {
...
     // Class under test
-    private lateinit var handler: StartRegistrationHandler
+    private lateinit var handler: RegistrationHandler<StartRegistration>
 
     // test dependencies
     lateinit var memberInfoFactory: MemberInfoFactory
-    lateinit var membershipGroupReader: MembershipGroupReader
-    lateinit var membershipGroupReaderProvider: MembershipGroupReaderProvider
     lateinit var membershipPersistenceClient: MembershipPersistenceClient
     lateinit var membershipQueryClient: MembershipQueryClient
```
Comment (nikinagy) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandlerTest.kt]: Make these private too.

File: components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/VerifyMemberHandlerTest.kt [hunk 8]
```diff
@@ -79,6 +88,40 @@ class VerifyMemberHandlerTest {
...
+
+        verify(membershipPersistenceClient, never()).setRegistrationRequestStatus(
+            any(),
+            any(),
+            any(),
+        )
+
+        assertThat(result.outputStates).hasSize(1)
+            .anyMatch {
+                (it.value as? RegistrationCommand)?.command is DeclineRegistration
```
Comment (nikinagy) [inline components/membership/registration-impl/src/test/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/VerifyMemberHandlerTest.kt]: Check reason.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt (hunk 2 above)
Comment (yift-r3) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt:58]: The `memberTypeChecker` was supposed to be that helper. It's easier to manage what happens if it's `false`/`true` than throwing an exception (which was my original plan).

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt [hunk 9]
```diff
@@ -82,6 +80,9 @@ class StartRegistrationHandler(
             }
 
             val mgmMemberInfo = getMGMMemberInfo(mgmHoldingId)
+            validateRegistrationRequest(!memberTypeChecker.isMgm(pendingMemberHoldingId)) {
```
Comment (nikinagy) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt]: We would need to move this check earlier before even persisting the registration request.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandler.kt [hunk 10]
```diff
@@ -70,64 +79,83 @@ internal class ApproveRegistrationHandler(
...
         val registrationId = state.registrationId
-        val persistState = membershipPersistenceClient.setMemberAndRegistrationRequestAsApproved(
-            viewOwningIdentity = approvedBy.toCorda(),
-            approvedMember = approvedMember.toCorda(),
-            registrationRequestId = registrationId,
-        )
-        val memberInfo = persistState.getOrThrow()
+        val messages = try {
+            val mgm = memberTypeChecker.getMgmMemberInfo(approvedBy.toCorda())
+                ?: throw CordaRuntimeException("Could not approve $registrationId - member ${approvedBy.x500Name} is an MGM.")
```
Comment (YashNabar) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandler.kt]: Shouldn't this be "is **not** an MGM"?

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandler.kt [hunk 11]
```diff
@@ -70,64 +79,83 @@ internal class ApproveRegistrationHandler(
...
+            if (memberTypeChecker.isMgm(approvedMember)) {
+                throw CordaRuntimeException("Could not approve $registrationId by ${approvedMember.x500Name} it is not an MGM.")
```
Comment (YashNabar) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandler.kt]: This might explain the situation better -
```suggestion
                throw CordaRuntimeException("The registration request: '$registrationId' cannot be approved by ${approvedMember.x500Name} as it is not an MGM.")
```

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/VerifyMemberHandler.kt [hunk 12]
```diff
@@ -3,41 +3,58 @@ package net.corda.membership.impl.registration.dynamic.handler.mgm
...
-        membershipPersistenceClient.setRegistrationRequestStatus(
-            mgm.toCorda(),
-            registrationId,
-            RegistrationStatus.PENDING_MEMBER_VERIFICATION
-        )
-        return RegistrationHandlerResult(
-            RegistrationState(registrationId, member, mgm),
+        val messages = try {
+            if (!memberTypeChecker.isMgm(mgm)) {
+                throw CordaRuntimeException("Could not verify $registrationId with ${mgm.x500Name} - Not an MGM.")
```
Comment (YashNabar) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/VerifyMemberHandler.kt]: Can we have something like " registration request: '$registrationId' " in all places where we're using these IDs? Having only the ID in the message may not be clear without the context

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/VerifyMemberHandler.kt [hunk 13]
```diff
@@ -48,6 +65,21 @@ class VerifyMemberHandler(
                     )
                 )
             )
+        } catch (e: Exception) {
+            logger.warn("Could not verify member fro registration $registrationId", e)
```
Comment (YashNabar) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/VerifyMemberHandler.kt]:
```suggestion
            logger.warn("Member verification failed for registration request: '$registrationId'.", e)
```

**Review by YashNabar — COMMENTED** [review]
Could we revise some of our error messages so that they include more context

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt [hunk 14]
```diff
@@ -15,39 +19,61 @@ import net.corda.membership.persistence.client.MembershipPersistenceClient
...
-            content = SetOwnRegistrationStatus(
+        val messages = try {
+            val success = command.verificationResponse.payload.items.firstOrNull {
+                it.key == VERIFIED
+            }?.value?.toBooleanStrictOrNull() ?: false
+            if (!success) {
+                val reasons = command.verificationResponse.payload.items
+                    .filter { it.key == FAILURE_REASONS }
+                    .map { it.value }
+                val message = "Could no verify registration request: '$registrationId' - $reasons"
```
Comment (nikinagy) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt]: This should be "Could not verify"

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt [hunk 15]
```diff
@@ -15,39 +19,61 @@ import net.corda.membership.persistence.client.MembershipPersistenceClient
...
+                throw CordaRuntimeException(message)
+            }
+            if (memberTypeChecker.isMgm(member)) {
+                throw CordaRuntimeException("Member ${member.x500Name} is an MGM and can not be register.")
```
Comment (nikinagy) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt]: this should be: "is an MGM and can not register."

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt [hunk 16]
```diff
@@ -74,6 +72,10 @@ class StartRegistrationHandler(
 
         val (outputCommand, outputStates) = try {
             logger.info("Persisting the received registration request.")
+            validateRegistrationRequest(!memberTypeChecker.isMgm(pendingMemberHoldingId)) {
+                "Registration request is registering a MGM holding identity."
```
Comment (nikinagy) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt]: "Registration request is registering an MGM holding identity."

**Review by nikinagy — COMMENTED** [review]
Only a few nit left.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/MemberTypeChecker.kt [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+internal class MemberTypeChecker(
+    private val membershipGroupReaderProvider: MembershipGroupReaderProvider,
+) {
+    fun isMgm(identity: DataHoldingIdentity) = isMgm(identity.toCorda())
+
+    fun isMgm(identity: CordaHoldingIdentity) =
+        getMgmMemberInfo(identity) != null
+
+    fun getMgmMemberInfo(identity: CordaHoldingIdentity): MemberInfo? {
```
Comment (charlieR3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/MemberTypeChecker.kt:18]: Is it right for this function to be publicly visible? Getting member info seems out of scope for a type checker class.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandler.kt [hunk 18]
```diff
@@ -70,64 +79,87 @@ internal class ApproveRegistrationHandler(
...
-        )
-        val memberInfo = persistState.getOrThrow()
+        val messages = try {
+            val mgm = memberTypeChecker.getMgmMemberInfo(approvedBy.toCorda())
+                ?: throw CordaRuntimeException(
+                    "Could not approve registration request: '$registrationId' - member ${approvedBy.x500Name} is not an MGM."
+                )
+            if (memberTypeChecker.isMgm(approvedMember)) {
+                throw CordaRuntimeException(
+                    "The registration request: '$registrationId' cannot be approved by ${approvedMember.x500Name} as it is an MGM."
```
Comment (charlieR3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ApproveRegistrationHandler.kt:89]:
```suggestion
                    "The registration request: '$registrationId' cannot be approved for ${approvedMember.x500Name} as it is an MGM."
```

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt [hunk 19]
```diff
@@ -15,39 +19,61 @@ import net.corda.membership.persistence.client.MembershipPersistenceClient
...
+                    .filter { it.key == FAILURE_REASONS }
+                    .map { it.value }
+                val message = "Could not verify registration request: '$registrationId' - $reasons"
+                throw CordaRuntimeException(message)
+            }
+            if (memberTypeChecker.isMgm(member)) {
+                throw CordaRuntimeException("Member ${member.x500Name} is an MGM and can not register.")
+            }
+            if (!memberTypeChecker.isMgm(mgm)) {
+                throw CordaRuntimeException("Member ${mgm.x500Name} is not an MGM and can not register.")
```
Comment (charlieR3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/ProcessMemberVerificationResponseHandler.kt]:
```suggestion
                throw CordaRuntimeException("Member ${mgm.x500Name} is not an MGM and can not process member's registration.")
```

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/MemberTypeChecker.kt (hunk 17 above)
Comment (yift-r3) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/MemberTypeChecker.kt:18]: The name might be wrong, but both the start and approve registration need the MGM info, and to check if the member is an MGM we need the member info, so I thought it would be better to remove the expose this function as part of the checker (the class is internal, so it's not entirely public).

**Review by charlieR3 — DISMISSED** [review]
LGTM

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt [hunk 20]
```diff
@@ -74,6 +72,10 @@ class StartRegistrationHandler(
 
         val (outputCommand, outputStates) = try {
             logger.info("Persisting the received registration request.")
```
Comment (nikinagy) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/StartRegistrationHandler.kt]: nit: I think this message should come after the validation.

**Review by nikinagy — DISMISSED** [review]
LGTM

**Review by charlieR3 — DISMISSED** [review]
LGTM

**Review by charlieR3 — APPROVED** [review]
LGTM
